### ISSUE TITLE
Difficutly populating nested data

### ISSUE THREAD

**Issue by jwerre** [caolan/highland#340]
I'm trying to 'join' some collection references from a MongoDB stream using Highland.js and I'm having some difficulty populating the id references with the actual document using `_.map`. Am I not using map correctly? Should I be using a different method altogether? Any help would be much appreciated.

	MongoClient = require('mongodb')
	_ = require('highland')

	MongoClient.connect 'mongodb://localhost/blog', function(err, db) {

		_( db.collection('posts').find(query).stream() )
			.each(function(post) {
				console.log(post)
				// {
				// 	_id: 557cbf50060f9e90521de109
				// 	title: "My Blog Post"
				// 	comments: [
				// 		"55a40877d804d87b433b2696",
				// 		"55a40894d804d87b433b2697",
				// 		"55a40894d804d87b433b2698",
				// 	]
				// }
				post.comments = _(post.comments).map(function(commentId) {
					_.wrapCallback( db.collection('comments').findOne )({_id: commentId}).apply(function(comment){
						return comment
					});
				});
				console.log(post)
			}).done(function() {
				console.log("THUNK!");
			});
	}
	

`post.comments` ends up a Higland stream object:

	{
		_id: 557cbf50060f9e90521de109
		title: "My Blog Post"
		comments: { 
			__HighlandStream__: true,
			...
		},

	}

...but I'm looking for something more along these lines:

	{
		_id: 557cbf50060f9e90521de109
		title: "My Blog Post"
		comments: [
			{ 
				_id "55a40877d804d87b433b2696"
				body: "Curabitur sit amet turpis eget."
			}
			{ 
				_id "55a40894d804d87b433b2697"
				body: "Cras ullamcorper magna urna, ut."
			}
			...
		]

	}

**Comment by quarterto** [comment]
`map` returns a stream. You're assigning that stream to `post.comments`. Instead of using `each`, `map` and `apply`, you should be using `flatMap`, and treating streams as opaque containers rather than things that you modify. I actually answered a very similar Stack Overflow question recently, which goes into more detail: http://stackoverflow.com/a/31314941/805227

**Comment by jwerre** [comment]
I initially tried using `flatMap`

    _( db.collection('posts').find(query).stream() )
        .flatMap(function(post){}) 

I get `Uncaught Error: Expected Stream, got undefined`

**Comment by quarterto** [comment]
The function you pass to `flatMap` *has* to return a Stream. `flatMap` runs once on each item in the stream, takes the streams you return, and returns another stream that will contain the results you the streams you returned, joined together. Because it encapsulates and waits for the return streams, you can use it to sequence asynchronous actions.

Here's a concrete example, using your example above, annotated with what's in the stream before and after each action:

```js
MongoClient = require('mongodb')
_ = require('highland')

MongoClient.connect 'mongodb://localhost/blog', function(err, db) {
    _( db.collection('posts').find(query).stream() ) // this is a Stream of Posts
        .flatMap(function(post) {
            return _(post.comments).flatMap(function(commentId) { // Stream of Ids
                var collection = db.collection('comments'); // need to bind
                return _.wrapCallback(collection.findOne.bind(collection))({_id: commentId}) // Stream of Comments
                  .collect() // Stream of Array of Comments
                  .map(function(comments){
                      post.comments = comments;
                      return post;
                  }); // Stream of single Post, with .comments as array of comments
            }); // Stream of Posts, with .comments as array of comments
        }).each(function(post) {
            // do something with the post
            console.log("THUNK!");
        });
}
```

It might seem tedious or unnecessary, but thinking about streams in terms of the types of object they contain, and `map` and `flatMap` as transforming between different types of streams (`map` by a single value, `flatMap` joining streams together) can be really helpful. It's also helpful to try and stay in Highland-land as long as possible, chaining together async operations with `flatMap`, and only causing a thunk when you really need the value outside the context of the stream (it's counterproductive to use e.g. `flatMap` to cause side-effects, as it almost never runs when you expect).

Hope this helps!

**Comment by jwerre** [comment]
Yes, that helps... the fog is lifting! One last question. Since you're returning the post for each comment the result duplicates the post for each individual comment. Is there a way to group the comments into a single array and return the post once?

**Comment by quarterto** [comment]
That's what the `collect` is supposed to do. It groups everything in the stream into an array and sends the array down the returned stream. Is that not what your seeing?

**Comment by jwerre** [comment]
No, it's unwinding the Post duplicating it for each comment and adding one comment per post. So instead of getting 1 post with 8 comments I get 8 posts with 1 comment each. I think this is because we're returning the post inside the comments map:

    return _.wrapCallback(collection.findOne.bind(collection))({_id: commentId}) // Stream of Comments
        .collect() // Stream of Array of Comments
        .map(function(comments){
            post.comments = comments;
            return post; // <------ HERE
        });

**Comment by quarterto** [comment]
Ah, my bad. The `.collect().map(...)` should be *outside* the `flatMap`. Currently it's `collect`ing a single comment, then `map`ping that comment, attaching it to the post, then returning that to `flatMap` which concats them all together. So, it should be:

```js
MongoClient = require('mongodb')
_ = require('highland')

MongoClient.connect 'mongodb://localhost/blog', function(err, db) {
    _( db.collection('posts').find(query).stream() ) // this is a Stream of Posts
        .flatMap(function(post) {
            return _(post.comments).flatMap(function(commentId) { // Stream of Ids
                var collection = db.collection('comments'); // need to bind
                return _.wrapCallback(collection.findOne.bind(collection))({_id: commentId}) // Stream of single Comment
            }) // Stream of Comments
              .collect() // Stream of Array of Comments
              .map(function(comments){
                  post.comments = comments;
                  return post;
              }); // Stream of single Post, with .comments as array of comments
           }); // Stream of Posts, with .comments as array of comments
        }) // Stream of Posts, with .comments as array of comments
        .each(function(post) {
            // do something with the post
            console.log("THUNK!");
        });
}
```

**Comment by jwerre** [comment]
That did it! Thanks for the primer.

**Comment by quarterto** [comment]
Happy to help. I think this makes a good basis for a "Highland cookbook". I might collect this (and that SO answer) into a blog post.

**Comment by jwerre** [comment]
I'll going to be working with this for the next couple of months making JSON, CSV, Excel, Word, PDFs files... I'll try and contribute the the cookbook along the way.

**Comment by saitheexplorer** [comment]
I'm trying to follow this issue to build a processor that reads documents from a Mongo database, adds a computed field, and updates the record back in the database.

In the example above, the connection is not being closed - I'm having trouble trying to close the database in the `.done()` callback, because the connection is closed before the updates finish.

Is my scenario an appropriate use case for Highland? I've been able to build scripts which stream from database, transform the records, and write to `stdout` or to other streams, but I'm struggling with performing async operations on the stream without either using a clunky `Stream.consume` implementation or bringing in `async` for the `queue` method. I'm fine with those approaches; they just seem like they're not the most efficient way of doing things.

**Comment by vqvu** [comment]
@saitheexplorer This is definitely something that Highland supports.

Can you create a new issue with some sample code that illustrates what you're trying to do? I can help you there. From your description, you shouldn't need to use `Stream.consume` or `async`. I can't tell from your description why the `.done()` callback is being executed early; I'll be able to tell you more if I see your code.

**Comment by saitheexplorer** [comment]
@vqvu Sure - I started building a simplified use case, and my code worked :) I'm going to compare the two versions and see where I'm going wrong.

On a tangentially related note - is there a Highland cookbook in the works? In building the use cases mentioned above, I wrote two blocks which achieved the same objective - once using `async` and once using Highland. Given the relative ubiquity of `async`, having a 'translation guide' may be helpful for people like me who are new to streams / Highland.
